### PR TITLE
Separate canonical data methods from ExerciseCase

### PR DESCRIPTION
**Pull request by Insti**
> ExerciseCase should not be an OpenStruct and the canonical data should be kept separate from the methods of this class.

from Generator Improvements exercism/xruby#485.

Supersedes #640 (Closed) 
Depends on #650 (Closed/Merged)
Completes #622 (Closed) to a usable state.

### CODE REVIEW

**Comment by Insti** [comment]
There are a lot of changes in https://github.com/exercism/xruby/pull/648/commits/ba6daa3dafbe4406091cd590c3b3ed0fc101a52c that are not directly related to separating the canonical data methods from ExerciseCase that I want to pull out of this PR.

**Comment by Insti** [comment]
#650 has been merged, rebased to master and force-pushed this branch.

File: test/generator/exercise_case_test.rb [hunk 1]
```diff
@@ -3,45 +3,41 @@
 module Generator
   class ExerciseCaseTest < Minitest::Test
     def test_name
-      assert_equal 'test_foo', ExerciseCase.new(description: 'foo').name
+      subject = ExerciseCase.new(canonical: OpenStruct.new('description' => 'foo'))
```
Comment (kotp) [inline test/generator/exercise_case_test.rb]: Why the change in style from `symbol: key` to `'string' => key` here, while maintaining the style of `symbol: key` in the most of the rest of the file?  (see line 34 as another deviation from the style)

File: exercises/beer-song/.meta/generator/beer_song_case.rb [hunk 2]
```diff
@@ -17,7 +17,7 @@ def beer_song_arguments
     if property == 'verse'
       number
     else
-      "%s, %s" % [self["beginning"], self["end"]]
+      "%s, %s" % [self.beginning, self.end]
```
Comment (Insti) [inline exercises/beer-song/.meta/generator/beer_song_case.rb]: These names come from the canonical data so we're constrained by what it provides.

File: test/generator/exercise_case_test.rb (hunk 1 above)
Comment (Insti) [inline test/generator/exercise_case_test.rb]: Fair comment. Updated to use symbols.

File: exercises/wordy/.meta/generator/wordy_case.rb [hunk 3]
```diff
@@ -16,7 +16,7 @@ def indent(size, lines)
   end
 
   def assertion
-    return error_assertion unless expected
+    return error_assertion if expected == false
```
Comment (kotp) [inline exercises/wordy/.meta/generator/wordy_case.rb]: `return error_assertion unless expected`

Or 

`return error_assertion if expected.false?`

Comment (kotp) [inline exercises/wordy/.meta/generator/wordy_case.rb]: How do the other cases address the assertions?

Comment (Insti) [inline exercises/wordy/.meta/generator/wordy_case.rb]: Where is `false?` defined?

Neither of these will work.

`expected` comes from the canonical data and is either not present or an integer or false.

Comment (kotp) [inline exercises/wordy/.meta/generator/wordy_case.rb]: That answers that then. :+1:   Thanks.
